### PR TITLE
PWGHF: Add Ω0c creator and selector

### PR DESCRIPTION
**Pull request by ZFederica**

### CODE REVIEW

File: PWGHF/TableProducer/candidateSelectorOmegac.cxx [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+  Configurable<double> V0RadiusMin{"V0RadiusMin", 2, "Min V0 radius"};
+  Configurable<double> V0CosPACut{"V0CosPACut", 0.95, "Cos PA V0 cut"};
+  Configurable<double> CascCosPACut{"CascCosPACut", 0.95, "Cos PA cascade cut"};
+  Configurable<double> dcaCascDauMax{"dcaCascDauMax", 5.0, "Max DCA cascade daughters"};
+  Configurable<double> dcaV0DauMax{"dcaV0DauMax", 5.0, "Max DCA V0 daughters"};
+  Configurable<double> dcaOmegacDauMax{"dcaOmegacDauMax", 5.0, "Max DCA omegac daughters"};
+
+  // limit charm baryon invariant mass spectrum
+  Configurable<double> LowerLimitSpectrum{"LowerLimitSpectrum", 2.4, "Lower limit invariant mass spectrum charm baryon"};
+  Configurable<double> UpperLimitSpectrum{"UpperLimitSpectrum", 3.0, "Upper limit invariant mass spectrum charm baryon"};
```
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorOmegac.cxx:47]: Use camelCase

File: PWGHF/TableProducer/candidateSelectorOmegac.cxx [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+struct HfCandidateSelectorOmegac {
+  Produces<aod::HFSelOmegacCandidate> hfSelOmegacCandidate;
+
+  // LF analysis selections
+  // zPV -> can be already set in HFeventselection -> 10 cm
+  // sel8 -> can be already set in HFeventselection -> true
+  Configurable<double> CascRadiusMin{"CascRadiusMin", 0.5, "Min cascade radius"};
+  Configurable<double> V0RadiusMin{"V0RadiusMin", 2, "Min V0 radius"};
+  Configurable<double> V0CosPACut{"V0CosPACut", 0.95, "Cos PA V0 cut"};
+  Configurable<double> CascCosPACut{"CascCosPACut", 0.95, "Cos PA cascade cut"};
```
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorOmegac.cxx:40]: Use `Min/Max` instead of `Cut`

File: PWGHF/TableProducer/candidateSelectorOmegac.cxx [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+  // limit charm baryon invariant mass spectrum
+  Configurable<double> LowerLimitSpectrum{"LowerLimitSpectrum", 2.4, "Lower limit invariant mass spectrum charm baryon"};
+  Configurable<double> UpperLimitSpectrum{"UpperLimitSpectrum", 3.0, "Upper limit invariant mass spectrum charm baryon"};
+
+  // kinematic selections
+  Configurable<double> etaTrackMax{"etaTrackMax", 0.8, "Max absolute value of eta"};
+  Configurable<double> ptMinPiFromCasc{"ptMinPiFromCasc", 0.15, "Min pT pi <- casc"};
+  Configurable<double> ptMinPiFromOme{"ptMinPiFromOme", 0.2, "Min pT pi <- omegac"};
+  Configurable<double> dcaxyPriPiMin{"dcaxyPriPiMin", 0., "Min dcaxy primary pi track to PV"};
+  Configurable<double> dcaxyPriPiMax{"dcaxyPriPiMax", 10., "Max dcaxy primary pi track to PV"};
```
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorOmegac.cxx:54]:
```suggestion
  Configurable<double> ptPiFromCascMin{"ptMinPiFromCasc", 0.15, "Min pT pi <- casc"};
  Configurable<double> ptPiFromOmeMin{"ptMinPiFromOme", 0.2, "Min pT pi <- omegac"};
  Configurable<double> dcaXYPriPiMin{"dcaxyPriPiMin", 0., "Min dcaxy primary pi track to PV"};
  Configurable<double> dcaXYPriPiMax{"dcaxyPriPiMax", 10., "Max dcaxy primary pi track to PV"};
```

File: PWGHF/TableProducer/candidateSelectorOmegac.cxx [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+
+  Configurable<double> ptCandMin{"ptCandMin", 0., "Lower bound of candidate pT"};
+  Configurable<double> ptCandMax{"ptCandMax", 50., "Upper bound of candidate pT"};
+
+  // PID options
+  Configurable<bool> PidTpcOnly{"PidTpcOnly", true, "Perform PID using only TPC"};
+  Configurable<bool> PidTpcTofCombined{"PidTpcTofCombined", false, "Perform PID using TPC & TOF"};
```
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorOmegac.cxx:61]:
```suggestion
  // PID options
  Configurable<bool> usePidTpcOnly{"PidTpcOnly", true, "Perform PID using only TPC"};
  Configurable<bool> usePidTpcTofCombined{"PidTpcTofCombined", false, "Perform PID using TPC & TOF"};
```

File: PWGHF/TableProducer/candidateSelectorOmegac.cxx [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+  Configurable<double> SigmaInvMassLambda{"SigmaInvMassLambda", 0.0025, "Invariant mass cut for lambda (sigma)"};
+  Configurable<double> SigmaInvMassCascade{"SigmaInvMassCascade", 0.0025, "Invariant mass cut for cascade (sigma)"};
+  Configurable<int> nSigmaInvMassCut{"nSigmaInvMassCut", 4, "Number of sigma for invariant mass cut"};
+
+  // detector clusters selections
+  Configurable<int> nClustersTpcMin{"nClustersTpcMin", 70, "Minimum number of TPC clusters requirement"};
+  Configurable<int> TpcCrossedRowsMin{"TpcCrossedRowsMin", 70, "Minimum number of TPC crossed rows requirement"};
+  Configurable<double> TpcCrossedRowsOverFindableClustersRatioMin{"TpcCrossedRowsOverFindableClustersRatioMin", 0.8, "Minimum ratio TPC crossed rows over findable clusters requirement"};
+  Configurable<int> nClustersItsMin{"nClustersItsMin", 3, "Minimum number of ITS clusters requirement for pi <- Omegac"};
+  Configurable<int> nClustersItsInnBarrMin{"nClustersItsInnBarrMin", 1, "Minimum number of ITS clusters in inner barrel requirement for pi <- Omegac"};
```
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorOmegac.cxx:85]: Use camelCase

File: PWGHF/TableProducer/candidateSelectorOmegac.cxx [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+
+  // PID options
+  Configurable<bool> PidTpcOnly{"PidTpcOnly", true, "Perform PID using only TPC"};
+  Configurable<bool> PidTpcTofCombined{"PidTpcTofCombined", false, "Perform PID using TPC & TOF"};
+
+  // PID - TPC selections
+  Configurable<double> ptMinPidTpc{"ptMinPidTpc", 0.15, "Lower bound of track pT for TPC PID"};
+  Configurable<double> ptMaxPidTpc{"ptMaxPidTpc", 5., "Upper bound of track pT for TPC PID"};
+  Configurable<double> nSigmaTpc{"nSigmaTpc", 3., "Nsigma cut on TPC only"};
+  Configurable<double> nSigmaTpcCombined{"nSigmaTpcCombined", 5., "Nsigma cut on TPC combined with TOF"};
```
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorOmegac.cxx:67]: Use `Min/Max` properly. See the other selectors.

File: PWGHF/TableProducer/candidateSelectorOmegac.cxx [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+        pidProton = selectorProton.getStatusTrackPIDTPC(trackPrFromLam);
+        pidPiFromLam = selectorPionFromV0.getStatusTrackPIDTPC(trackPiFromLam);
+        pidPiFromCasc = selectorPionFromCasc.getStatusTrackPIDTPC(trackPiFromCasc);
+        pidPiFromOme = selectorPionFromOme.getStatusTrackPIDTPC(trackPiFromOmeg);
+      } else if (PidTpcTofCombined) {
+        pidProton = selectorProton.getStatusTrackPIDAll(trackPrFromLam);
+        pidPiFromLam = selectorPionFromV0.getStatusTrackPIDAll(trackPiFromLam);
+        pidPiFromCasc = selectorPionFromCasc.getStatusTrackPIDAll(trackPiFromCasc);
+        pidPiFromOme = selectorPionFromOme.getStatusTrackPIDAll(trackPiFromOmeg);
+      }
```
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorOmegac.cxx:270]: You should take care of the possibility that `PidTpcOnly` and `PidTpcTofCombined` might be both `true` or both `false`.

File: PWGHF/TableProducer/candidateSelectorOmegac.cxx [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+      double invMassLambda = 0;
+      if (signDecay < 0) {
+        invMassLambda = candidate.invMassLambda();
+      } else if (signDecay > 0) {
+        invMassLambda = candidate.invMassAntiLambda();
+      }
+      double invMassCascade = candidate.invMassCascade();
+      double invMassOmegac = candidate.invMassOmegac();
+
+      if (std::abs(invMassLambda - 1.11568) < (nSigmaInvMassCut * SigmaInvMassLambda)) {
```
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorOmegac.cxx:305]: Do not use hard-coded constants.

File: PWGHF/TableProducer/candidateSelectorOmegac.cxx [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+        statusInvMassLambda = 1;
+        if (statusPidLambda == 1 && statusPidCascade == 1 && statusPidOmegac == 1) {
+          hTest2->Fill(3.5);
+        }
+      }
+
+      if (std::abs(invMassCascade - 1.32171) < (nSigmaInvMassCut * SigmaInvMassCascade)) {
```
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorOmegac.cxx:312]: Do not use hard-coded constants.

File: PWGHF/TableProducer/candidateSelectorOmegac.cxx [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+        hInvMassOmegacNotFixed->Fill(candidate.massOmegacNotFixed());
+      }
+    }
+  }
+};
+
+WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
+{
+  return WorkflowSpec{
+    adaptAnalysisTask<HfCandidateSelectorOmegac>(cfgc, TaskName{"hf-omegac-candidate-selector"})};
```
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorOmegac.cxx:379]: Do not use `TaskName`.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/DataModel/CandidateReconstructionTables.h [hunk 11]
```diff
@@ -1182,6 +1182,146 @@ DECLARE_SOA_TABLE(HfCandXiccMcGen, "AOD", "HFCANDXICCMCGEN", //!
...
+DECLARE_SOA_COLUMN(DebugMCRec, debugMCRec, int8_t);         // debug flag for mis-association reconstruction level
+DECLARE_SOA_COLUMN(FlagMCMatchGen, flagMCMatchGen, int8_t); // generator level
+
+// mapping of decay types
+enum DecayType { OmegacToXiPi = 0 };
+
+} // end of namespace hf_cand_omegac
+
+// declare dedicated Omegac candidate table
+DECLARE_SOA_TABLE(HfCandOmegacBase, "AOD", "HFCANDOMECBASE",
```
Comment (vkucera) [inline PWGHF/DataModel/CandidateReconstructionTables.h:1285]: Is there a reason to call it `Base` if you don't extend it?

File: PWGHF/TableProducer/candidateCreatorOmegac.cxx [hunk 12]
```diff
@@ -0,0 +1,516 @@
...
+        }
+
+        const std::array<float, 2> arrMassCascade = {1.11568, 0.13957};
+        double mCascade = RecoDecay::m(array{pvecV0_d, pvecPionFromCasc}, arrMassCascade);
+        double mCascadeNotFixed = RecoDecay::m(array{pvecV0_d, pvecPionFromCasc}, array{my_mLambda, 0.13957});
+        double mCascLF = casc.mXi();
+
+        const std::array<float, 2> arrMassOmegac = {1.32171, 0.13957};
+        double mOmegac = RecoDecay::m(array{pvecCasc_d, pvecPionFromOmegac}, arrMassOmegac);
+        double mOmegacNotFixed = RecoDecay::m(array{pvecCasc_d, pvecPionFromOmegac}, array{mCascadeNotFixed, 0.13957});
```
Comment (vkucera) [inline PWGHF/TableProducer/candidateCreatorOmegac.cxx:347]: Do not use hard-coded constants.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/DataModel/CandidateReconstructionTables.h [hunk 13]
```diff
@@ -1182,6 +1182,146 @@ DECLARE_SOA_TABLE(HfCandXiccMcGen, "AOD", "HFCANDXICCMCGEN", //!
...
+DECLARE_SOA_COLUMN(CTauCascade, ctauCascade, double);
+DECLARE_SOA_COLUMN(CTauV0, ctauV0, double);
+DECLARE_SOA_COLUMN(EtaV0PosDau, etaV0PosDau, double);
+DECLARE_SOA_COLUMN(EtaV0NegDau, etaV0NegDau, double);
+DECLARE_SOA_COLUMN(EtaPiFromCasc, etaPiFromCasc, double);
+DECLARE_SOA_COLUMN(EtaPiFromOme, etaPiFromOme, double);
+DECLARE_SOA_COLUMN(EtaOmegac, etaOmegac, double);
+DECLARE_SOA_COLUMN(EtaCascade, etaCascade, double);
+DECLARE_SOA_COLUMN(EtaV0, etaV0, double);
+DECLARE_SOA_COLUMN(MyMassLambda, myMassLambda, double);
```
[the remark this answers is not in the document]
Comment (vkucera) [reply PWGHF/DataModel/CandidateReconstructionTables.h:1259]: Yes, that would be useful.

File: PWGHF/TableProducer/candidateCreatorOmegac.cxx [hunk 14]
```diff
@@ -0,0 +1,516 @@
...
+
+      // info from LF table
+      array<float, 3> pvecV0LFTable = {v0Element.px(), v0Element.py(), v0Element.pz()};
+      array<float, 3> vertexV0 = {v0Element.x(), v0Element.y(), v0Element.z()};
+
+      // reconstruct V0 with DCAFitter
+      int nfv = dfv.process(trackParCovV0Dau0, trackParCovV0Dau1);
+      if (nfv == 0) {
+        continue;
+      }
```
[the remark this answers is not in the document]
Comment (ZFederica) [reply PWGHF/TableProducer/candidateCreatorOmegac.cxx:196]: I prefer to keep it, but I changed the name to make it more clear

File: PWGHF/TableProducer/candidateCreatorOmegac.cxx [hunk 15]
```diff
@@ -0,0 +1,516 @@
...
+      auto vertexV0FromFitter = dfv.getPCACandidate();
+      auto chi2PCAV0 = dfv.getChi2AtPCACandidate();
+      array<float, 3> pvecV0Dau0;
+      array<float, 3> pvecV0Dau1;
```
[the remark this answers is not in the document]
Comment (ZFederica) [reply PWGHF/TableProducer/candidateCreatorOmegac.cxx:200]: I know that the camelCase rule, but this was meant to be this way. pvec stands for pvector, it is meant as a whole object not as separate entities in a string name.

File: PWGHF/TableProducer/candidateCreatorOmegac.cxx [hunk 16]
```diff
@@ -0,0 +1,525 @@
...
+    dfv.setMinRelChi2Change(minRelChi2Change);
+    dfv.setUseAbsDCA(useAbsDCA);
+    dfv.setWeightedFinalPCA(useWeightedPCA);
+    dfv.setRefitWithMatCorr(refitWithMatCorr);
+
+    double massPionTrueValue = RecoDecay::getMassPDG(kPiPlus);    // pdg code 211
+    double massProtonTrueValue = RecoDecay::getMassPDG(kProton);  // pdg code 2212
+    double massLambdaTrueValue = RecoDecay::getMassPDG(kLambda0); // pdg code 3122
+    double massXiTrueValue = RecoDecay::getMassPDG(kXiMinus);     // pdg code 3312
+    double massOmegacTrueValue = RecoDecay::getMassPDG(kOmegac0); // pdg code 4332
```
Comment (vkucera) [inline PWGHF/TableProducer/candidateCreatorOmegac.cxx:152]: Please remove the `TrueValue` suffixes. They serve no purpose. There is no such a thing as true or false values of a particle mass.

File: PWGHF/TableProducer/candidateSelectorOmegac.cxx [hunk 17]
```diff
@@ -0,0 +1,387 @@
...
+    TrackSelectorPID selectorProton(kProton);
+    selectorProton.setRangePtTPC(ptPidTpcMin, ptPidTpcMax);
+    selectorProton.setRangeNSigmaTPC(-nSigmaTpcMax, nSigmaTpcMax);
+    selectorProton.setRangeNSigmaTPCCondTOF(-nSigmaTpcCombinedMax, nSigmaTpcCombinedMax);
+    selectorProton.setRangePtTOF(ptPidTofMin, ptPidTofMax);
+    selectorProton.setRangeNSigmaTOF(-nSigmaTofMax, nSigmaTofMax);
+    selectorProton.setRangeNSigmaTOFCondTPC(-nSigmaTofCombinedMax, nSigmaTofCombinedMax);
+
+    double massLambdaTrueValue = RecoDecay::getMassPDG(kLambda0);
+    double massXiTrueValue = RecoDecay::getMassPDG(kXiMinus);
```
Comment (vkucera) [inline PWGHF/TableProducer/candidateSelectorOmegac.cxx:135]: Remove `TrueValue`.

**Review by vkucera — CHANGES_REQUESTED** [review]

**Review by vkucera — APPROVED** [review]

**Review by vkucera — APPROVED** [review]

**Pull request merged**